v1/charm-dev.yaml

```diff
@@ -0,0 +1,99 @@
+description: A development and testing environment for charmers
+version: latest
+
+instances:
+  charm-dev:
+    image: 20.04
+    limits:
+      min-cpu: 2
+      min-mem: 4G
+      min-disk: 30G
+    timeout: 900
+    cloud-init:
+      vendor-data: |
+        package_update: true
+
+        packages:
+        - python3-pip
+        - jq
+        - gh
+        - sysstat
+        - zsh
+        - fzf
+        - tox
+
+        snap:
+          commands:
+          - snap install --classic juju
+          - snap install --classic microk8s
+          - snap alias microk8s.kubectl kubectl
+          - snap install --classic charmcraft
+          - snap install yq
+          - snap refresh
+
+        runcmd:
+        - DEBIAN_FRONTEND=noninteractive apt -y upgrade
+
+        - |
+          # disable swap
+          sysctl -w vm.swappiness=0
+          echo "vm.swappiness = 0" | tee -a /etc/sysctl.conf
+          swapoff -a
+
+        - |
+          # disable unnecessary services
+          systemctl disable man-db.timer man-db.service --now
+          systemctl disable apport.service apport-autoreport.service  --now
+          systemctl disable apt-daily.service apt-daily.timer --now
+          systemctl disable apt-daily-upgrade.service apt-daily-upgrade.timer --now
+          systemctl disable unattended-upgrades.service --now
+          systemctl disable motd-news.service motd-news.timer --now
+          systemctl disable bluetooth.target --now
+          systemctl disable ua-messaging.service ua-messaging.timer --now
+          systemctl disable ua-timer.timer ua-timer.service --now
+          systemctl disable systemd-tmpfiles-clean.timer --now
+
+        - |
+          # various tools via apt
+          # gh cli
+          curl -fsSL https://cli.github.com/packages/githubcli-archive-keyring.gpg | sudo gpg --dearmor -o /usr/share/keyrings/githubcli-archive-keyring.gpg
+          echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | sudo tee /etc/apt/sources.list.d/github-cli.list > /dev/null
+
+        - |
+          # apt cleanup
+          apt-get autoremove -y
+
+        - |
+          # oh-my-zsh + juju plugin
+          sudo -u ubuntu sh -c "$(curl -fsSL https://raw.githubusercontent.com/ohmyzsh/ohmyzsh/master/tools/install.sh)" "" --unattended
```

Comment:
This doesn't seem to have worked:

```plain
Zsh is not installed. Please install zsh first.
```

Reply:
Odd. It is listed under `packages:` and I do recall having it.
Will test again.

Reply:
So here's the culprit:

```plain
E: Unable to locate package gh
2022-01-13 14:05:18,163 - util.py[WARNING]: Failed to install packages: ['python3-pip', 'jq', 'gh', 'sysstat', 'zsh', 'fzf', 'tox']
```

It tries to install `gh` before its repository is added.

You could try using the [apt-configure module](https://cloudinit.readthedocs.io/en/latest/topics/modules.html#apt-configure) or simply install that package in `runcmd` after adding the repo.